File: apps/windows/ClipPlus.Windows/Update/GitHubReleaseClient.cs
namespace ClipPlus.Windows.Update;

using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

public sealed class GitHubReleaseClient
{
    public static readonly Uri LatestReleaseUri = new("https://api.github.com/repos/nicelnicel/ClipPlus/releases/latest");

    private readonly HttpClient httpClient;

    public GitHubReleaseClient(HttpClient? httpClient = null)
    {
        this.httpClient = httpClient ?? new HttpClient();
    }

    public static GitHubRelease DecodeRelease(string json)
    {
        return JsonSerializer.Deserialize<GitHubRelease>(json)
            ?? throw new UpdateException(UpdateErrorKind.InvalidVersion);
    }

    public static SelectedUpdateAsset SelectWindowsAsset(GitHubRelease release, UpdateVersion currentVersion)
    {
        if (release.Draft || release.Prerelease || !UpdateVersion.TryParse(release.TagName, out var releaseVersion))
        {
            throw new UpdateException(UpdateErrorKind.InvalidVersion);
        }

        if (releaseVersion.CompareTo(currentVersion) <= 0)
        {
            throw new UpdateException(UpdateErrorKind.UpToDate);
        }

        var asset = release.Assets.FirstOrDefault(asset => asset.Name == "ClipPlus-Windows-x64-full.exe")
            ?? throw new UpdateException(UpdateErrorKind.MissingAsset);
        var digest = NormalizedSha256Digest(asset.Digest);
        return new SelectedUpdateAsset(
            releaseVersion,
            asset.Name,
            asset.BrowserDownloadUrl,
            digest,
            asset.Size
        );
    }

    public async Task<GitHubRelease> FetchLatestReleaseAsync(CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, LatestReleaseUri);
        request.Headers.UserAgent.ParseAdd("ClipPlus");
        request.Headers.Accept.ParseAdd("application/vnd.github+json");
        using var response = await httpClient.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new UpdateException(UpdateErrorKind.DownloadFailed);
        }

        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        return DecodeRelease(json);
    }

    private static string NormalizedSha256Digest(string? digest)
    {
        if (string.IsNullOrWhiteSpace(digest))
        {
            throw new UpdateException(UpdateErrorKind.MissingDigest);
        }

        var normalized = digest.Trim().ToLowerInvariant();
        if (normalized.StartsWith("sha256:", StringComparison.Ordinal))
        {
            normalized = normalized["sha256:".Length..];
        }

        if (!Regex.IsMatch(normalized, "^[0-9a-f]{64}$"))
        {
            throw new UpdateException(UpdateErrorKind.InvalidDigest);
        }

        return normalized;
    }
}

public sealed record GitHubRelease(
    [property: JsonPropertyName("tag_name")] string TagName,
    [property: JsonPropertyName("draft")] bool Draft,
    [property: JsonPropertyName("prerelease")] bool Prerelease,
    [property: JsonPropertyName("assets")] IReadOnlyList<GitHubReleaseAsset> Assets
);

public sealed record GitHubReleaseAsset(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("browser_download_url")] Uri BrowserDownloadUrl,
    [property: JsonPropertyName("digest")] string? Digest,
    [property: JsonPropertyName("size")] long Size
);

File: apps/mac/Sources/ClipPlusMac/Update/GitHubReleaseClient.swift
import Foundation

struct GitHubReleaseClient {
    static let latestReleaseURL = URL(string: "https://api.github.com/repos/nicelnicel/ClipPlus/releases/latest")!

    static func decodeRelease(from data: Data) throws -> GitHubRelease {
        try JSONDecoder().decode(GitHubRelease.self, from: data)
    }

    static func selectMacAsset(
        from release: GitHubRelease,
        currentVersion: UpdateVersion
    ) throws -> SelectedUpdateAsset {
        guard !release.draft, !release.prerelease,
              let releaseVersion = UpdateVersion(release.tagName) else {
            throw UpdateError.invalidVersion
        }

        guard releaseVersion > currentVersion else {
            throw UpdateError.upToDate
        }

        guard let asset = release.assets.first(where: { $0.name == "ClipPlus-macOS.dmg" }) else {
            throw UpdateError.missingAsset
        }

        let digest = try normalizedSha256Digest(asset.digest)
        return SelectedUpdateAsset(
            version: releaseVersion,
            name: asset.name,
            downloadURL: asset.browserDownloadURL,
            sha256Hex: digest,
            size: asset.size
        )
    }

    func fetchLatestRelease() async throws -> GitHubRelease {
        var request = URLRequest(url: Self.latestReleaseURL)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("ClipPlus", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200..<300).contains(httpResponse.statusCode) else {
            throw UpdateError.downloadFailed
        }

        return try Self.decodeRelease(from: data)
    }

    private static func normalizedSha256Digest(_ digest: String?) throws -> String {
        guard let digest, !digest.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw UpdateError.missingDigest
        }

        let normalized = digest
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: "sha256:", with: "")
        guard normalized.range(of: #"^[0-9a-f]{64}$"#, options: .regularExpression) != nil else {
            throw UpdateError.invalidDigest
        }

        return normalized
    }
}
